File: flynn.rb
require 'websocket-client-simple'
require 'json'
require 'fileutils'
require 'readline'
require 'httparty'

$socket = nil
$ping_interval = nil
$countdown_interval = nil
$potential_points = 0
$countdown = "Calculating..."
$points_total = 0
$points_today = 0

def get_local_storage
  begin
    data = File.read('localStorage.json')
    JSON.parse(data)
  rescue
    {}
  end
end

def set_local_storage(data)
  current_data = get_local_storage
  new_data = current_data.merge(data)
  File.write('localStorage.json', JSON.generate(new_data))
end

def connect_websocket(user_id)
  return if $socket
  version = "v0.2"
  url = "wss://secure.ws.teneo.pro"
  ws_url = "#{url}/websocket?userId=#{URI.encode_www_form_component(user_id)}&version=#{URI.encode_www_form_component(version)}"
  $socket = WebSocket::Client::Simple.connect(ws_url)

  $socket.on :open do
    connection_time = Time.now.iso8601
    set_local_storage({ 'lastUpdated' => connection_time })
    puts "WebSocket connected at #{connection_time}"
    start_pinging
    start_countdown_and_points
  end

  $socket.on :message do |event|
    data = JSON.parse(event.data)
    puts "Received message from WebSocket: #{data}"
    if data['pointsTotal'] && data['pointsToday']
      last_updated = Time.now.iso8601
      set_local_storage({
        'lastUpdated' => last_updated,
        'pointsTotal' => data['pointsTotal'],
        'pointsToday' => data['pointsToday']
      })
      $points_total = data['pointsTotal']
      $points_today = data['pointsToday']
    end
  end

  $socket.on :close do
    $socket = nil
    puts "WebSocket disconnected"
    stop_pinging
  end

  $socket.on :error do |error|
    puts "WebSocket error: #{error}"
  end
end

def disconnect_websocket
  if $socket
    $socket.close
    $socket = nil
    stop_pinging
  end
end

def start_pinging
  stop_pinging
  $ping_interval = Thread.new do
    loop do
      if $socket && $socket.open?
        $socket.send(JSON.generate({ type: "PING" }))
        set_local_storage({ 'lastPingDate' => Time.now.iso8601 })
      end
      sleep 10
    end
  end
end

def stop_pinging
  $ping_interval.kill if $ping_interval
  $ping_interval = nil
end

trap('INT') do
  puts 'Received SIGINT. Stopping pinging...'
  stop_pinging
  disconnect_websocket
  exit 0
end

def start_countdown_and_points
  $countdown_interval&.kill
  update_countdown_and_points
  $countdown_interval = Thread.new do
    loop do
      update_countdown_and_points
      sleep 1
    end
  end
end

def update_countdown_and_points
  local_storage = get_local_storage
  last_updated = local_storage['lastUpdated']
  if last_updated
    next_heartbeat = Time.parse(last_updated) + (15 * 60)
    now = Time.now
    diff = next_heartbeat - now

    if diff > 0
      minutes = (diff / 60).to_i
      seconds = (diff % 60).to_i
      $countdown = "#{minutes}m #{seconds}s"

      max_points = 25
      time_elapsed = now - Time.parse(last_updated)
      time_elapsed_minutes = time_elapsed / 60
      new_points = [max_points, (time_elapsed_minutes / 15) * max_points].min
      new_points = new_points.round(2)

      if rand < 0.1
        bonus = rand * 2
        new_points = [max_points, new_points + bonus].min
        new_points = new_points.round(2)
      end

      $potential_points = new_points
    else
      $countdown = "Calculating..."
      $potential_points = 25
    end
  else
    $countdown = "Calculating..."
    $potential_points = 0
  end

  set_local_storage({ 'potentialPoints' => $potential_points, 'countdown' => $countdown })
end

def get_user_id
  login_url = "https://ikknngrgxuxgjhplbpey.supabase.co/auth/v1/token?grant_type=password"
  authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"
  apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag"

  email = Readline.readline('Email: ', true)
  password = Readline.readline('Password: ', true)

  begin
    response = HTTParty.post(login_url, 
      body: { email: email, password: password }.to_json,
      headers: {
        'Authorization' => authorization,
        'apikey' => apikey,
        'Content-Type' => 'application/json'
      }
    )

    user_id = response['user']['id']
    puts "User ID: #{user_id}"

    profile_url = "https://ikknngrgxuxgjhplbpey.supabase.co/rest/v1/profiles?select=personal_code&id=eq.#{user_id}"
    profile_response = HTTParty.get(profile_url,
      headers: {
        'Authorization' => authorization,
        'apikey' => apikey
      }
    )

    puts "Profile Data: #{profile_response.body}"
    set_local_storage({ 'userId' => user_id })
    start_countdown_and_points
    connect_websocket(user_id)
  rescue => e
    puts "Error: #{e.message}"
  end
end

def main
  local_storage_data = get_local_storage
  user_id = local_storage_data['userId']

  if !user_id
    user_id = Readline.readline('Please enter your user ID: ', true)
    set_local_storage({ 'userId' => user_id })
    start_countdown_and_points
    connect_websocket(user_id)
  else
    option = Readline.readline("Menu:\n1. Logout\n2. Start Running Node\n3. Get User ID\nChoose an option: ", true)
    case option
    when '1'
      set_local_storage({})
      puts 'Logged out successfully.'
      exit 0
    when '2'
      start_countdown_and_points
      connect_websocket(user_id)
    when '3'
      get_user_id
    else
      puts 'Invalid option. Exiting...'
      exit 0
    end
  end
end

main

File: flynn.php
<?php

require 'vendor/autoload.php';

use Ratchet\Client\WebSocket;
use Ratchet\Client\Connector;
use React\EventLoop\Loop;
use GuzzleHttp\Client;

$socket = null;
$pingInterval = null;
$countdownInterval = null;
$potentialPoints = 0;
$countdown = "Calculating...";
$pointsTotal = 0;
$pointsToday = 0;

function getLocalStorage() {
    try {
        $data = file_get_contents('localStorage.json');
        return json_decode($data, true);
    } catch (Exception $error) {
        return [];
    }
}

function setLocalStorage($data) {
    $currentData = getLocalStorage();
    $newData = array_merge($currentData, $data);
    file_put_contents('localStorage.json', json_encode($newData));
}

function connectWebSocket($userId) {
    global $socket;
    if ($socket) return;
    $version = "v0.2";
    $url = "wss://secure.ws.teneo.pro";
    $wsUrl = "{$url}/websocket?userId=" . urlencode($userId) . "&version=" . urlencode($version);

    $loop = Loop::get();
    $reactConnector = new Connector($loop);
    $reactConnector($wsUrl)->then(function(WebSocket $conn) use ($loop) {
        global $socket;
        $socket = $conn;

        $connectionTime = date('c');
        setLocalStorage(['lastUpdated' => $connectionTime]);
        echo "WebSocket connected at " . $connectionTime . "\n";
        startPinging($loop);
        startCountdownAndPoints($loop);

        $conn->on('message', function($msg) {
            $data = json_decode($msg, true);
            echo "Received message from WebSocket: " . print_r($data, true) . "\n";
            if (isset($data['pointsTotal']) && isset($data['pointsToday'])) {
                $lastUpdated = date('c');
                setLocalStorage([
                    'lastUpdated' => $lastUpdated,
                    'pointsTotal' => $data['pointsTotal'],
                    'pointsToday' => $data['pointsToday'],
                ]);
                global $pointsTotal, $pointsToday;
                $pointsTotal = $data['pointsTotal'];
                $pointsToday = $data['pointsToday'];
            }
        });

        $conn->on('close', function() {
            global $socket;
            $socket = null;
            echo "WebSocket disconnected\n";
            stopPinging();
        });
    }, function(Exception $e) {
        echo "WebSocket error: " . $e->getMessage() . "\n";
    });
}

function disconnectWebSocket() {
    global $socket;
    if ($socket) {
        $socket->close();
        $socket = null;
        stopPinging();
    }
}

function startPinging($loop) {
    global $pingInterval, $socket;
    stopPinging();
    $pingInterval = $loop->addPeriodicTimer(10, function() use ($socket) {
        if ($socket) {
            $socket->send(json_encode(['type' => 'PING']));
            setLocalStorage(['lastPingDate' => date('c')]);
        }
    });
}

function stopPinging() {
    global $pingInterval;
    if ($pingInterval) {
        Loop::get()->cancelTimer($pingInterval);
        $pingInterval = null;
    }
}

function startCountdownAndPoints($loop) {
    global $countdownInterval;
    if ($countdownInterval) {
        $loop->cancelTimer($countdownInterval);
    }
    updateCountdownAndPoints();
    $countdownInterval = $loop->addPeriodicTimer(1, 'updateCountdownAndPoints');
}

function updateCountdownAndPoints() {
    global $countdown, $potentialPoints;
    $localStorage = getLocalStorage();
    $lastUpdated = $localStorage['lastUpdated'] ?? null;
    if ($lastUpdated) {
        $nextHeartbeat = new DateTime($lastUpdated);
        $nextHeartbeat->modify('+15 minutes');
        $now = new DateTime();
        $diff = $nextHeartbeat->getTimestamp() - $now->getTimestamp();

        if ($diff > 0) {
            $minutes = floor($diff / 60);
            $seconds = $diff % 60;
            $countdown = "{$minutes}m {$seconds}s";

            $maxPoints = 25;
            $timeElapsed = $now->getTimestamp() - (new DateTime($lastUpdated))->getTimestamp();
            $timeElapsedMinutes = $timeElapsed / 60;
            $newPoints = min($maxPoints, ($timeElapsedMinutes / 15) * $maxPoints);
            $newPoints = round($newPoints, 2);

            if (mt_rand() / mt_getrandmax() < 0.1) {
                $bonus = mt_rand() / mt_getrandmax() * 2;
                $newPoints = min($maxPoints, $newPoints + $bonus);
                $newPoints = round($newPoints, 2);
            }

            $potentialPoints = $newPoints;
        } else {
            $countdown = "Calculating...";
            $potentialPoints = 25;
        }
    } else {
        $countdown = "Calculating...";
        $potentialPoints = 0;
    }

    setLocalStorage(['potentialPoints' => $potentialPoints, 'countdown' => $countdown]);
}

function getUserId() {
    $loginUrl = "https://ikknngrgxuxgjhplbpey.supabase.co/auth/v1/token?grant_type=password";
    $authorization = "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";
    $apikey = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imlra25uZ3JneHV4Z2pocGxicGV5Iiwicm9sZSI6ImFub24iLCJpYXQiOjE3MjU0MzgxNTAsImV4cCI6MjA0MTAxNDE1MH0.DRAvf8nH1ojnJBc3rD_Nw6t1AV8X_g6gmY_HByG2Mag";

    echo "Email: ";
    $email = trim(fgets(STDIN));
    echo "Password: ";
    $password = trim(fgets(STDIN));

    $client = new Client();
    try {
        $response = $client->post($loginUrl, [
            'json' => [
                'email' => $email,
                'password' => $password
            ],
            'headers' => [
                'Authorization' => $authorization,
                'apikey' => $apikey
            ]
        ]);

        $data = json_decode($response->getBody(), true);
        $userId = $data['user']['id'];
        echo "User ID: " . $userId . "\n";

        $profileUrl = "https://ikknngrgxuxgjhplbpey.supabase.co/rest/v1/profiles?select=personal_code&id=eq.{$userId}";
        $profileResponse = $client->get($profileUrl, [
            'headers' => [
                'Authorization' => $authorization,
                'apikey' => $apikey
            ]
        ]);

        $profileData = json_decode($profileResponse->getBody(), true);
        echo "Profile Data: " . print_r($profileData, true) . "\n";
        setLocalStorage(['userId' => $userId]);
        startCountdownAndPoints(Loop::get());
        connectWebSocket($userId);
    } catch (Exception $error) {
        echo "Error: " . $error->getMessage() . "\n";
    }
}

function main() {
    $localStorageData = getLocalStorage();
    $userId = $localStorageData['userId'] ?? null;

    if (!$userId) {
        echo "Please enter your user ID: ";
        $userId = trim(fgets(STDIN));
        setLocalStorage(['userId' => $userId]);
        startCountdownAndPoints(Loop::get());
        connectWebSocket($userId);
    } else {
        echo "Menu:\n1. Logout\n2. Start Running Node\n3. Get User ID\nChoose an option: ";
        $option = trim(fgets(STDIN));
        switch ($option) {
            case '1':
                setLocalStorage([]);
                echo "Logged out successfully.\n";
                exit(0);
                break;
            case '2':
                startCountdownAndPoints(Loop::get());
                connectWebSocket($userId);
                break;
            case '3':
                getUserId();
                break;
            default:
                echo "Invalid option. Exiting...\n";
                exit(0);
        }
    }
}

main();
Loop::get()->run();

